File: python/data_structures/priority_q.py
class PriorityQueue:
	class Node:
		def __init__(self, val, priority):
			self.value = val
			self.priority=priority
		def __gt__(self, other):
			return self.priority > other.priority

		def __lt__(self, other):
			return not self.__gt__(other)

		def __eq__(self, other):
			return self.priority == other.priority

		def __ge__(self, other):
			return self.__gt__(other) or self.__eq__(other)

	def __init__(self):
		self._values = []
		self._size = 0

	def __str__(self):
		return { str(p): v for p,v in self._values }

	def __repr__(self):
		tmp = ''
		for i,v in enumerate(self._values):
			tmp += f'{v.val}:{v.priority}'
			if i < self._size:
				tmp += ', '
		return tmp

	def shift_up(self):
		idx = self._size - 1
		while idx > 0:
			parent = (idx -1) >> 1
			if self._values[idx] >= self._values[parent]: break
			self._values[parent], self._values[idx] = self._values[idx], self._values[parent]
			idx = parent

	def enqueue(self, value, priority):
		self._values.append(self.Node(value, priority))
		self._size += 1
		self.shift_up()


	def dequeue(self):
		highest = None
		if self._size > 0:
			highest = self._values[0]
			end = self._values.pop()
			self._values[0] = end
			self._size -= 1
			self.shift_down()

		return highest

	def shift_down(self):
		el = self._values[0]
		idx = 0
		heap_size = self._size
		while idx < heap_size:
			swap_id = idx
			left = (idx << 1) + 1
			right = left+1
			if left < self._size:
				if self._values[left] < el:
					swap_id = left
			if right < self._size and self._values[swap_id] < self._values[right]:
						swap_id = right

			if swap_id == idx: break

			self._values[idx], self._values[swap_id] = self._values[swap_id], el
			idx = swap_id

File: php/data_structures/priority_q.php
<?php

class PQNode{
	public ?int $priority;
	public $value;
	public function __construct($value, $priority=null){
		$this->value = $value;
		$this->priority = $priority;
	}
	public function __toString(): string{
		return "{ value: $this->value, priority:$this->priority }";
	}
}

class PriorityQueue {
	private $values;
	private $len;
	public function __construct(){
		$this->values = array();
		$this->len = 0;
	}
	private function shift_up(){
		$index = $this->len - 1;
		while($index > 0){
			$parent_index = ($index -1) >> 1;
			if($this->values[$index] >= $this->values[$parent_index]) break;
			$tmp = $this->values[$parent_index];
			$this->values[$parent_index] = $this->values[$index];
			$this->values[$index] = $tmp;
			$index = $parent_index;
		}
	}
	private function shift_down(){
		$index = 0;
		while(true){
			$swap_id = $index;
			$left = ($index << 1) + 1;
			$right = $left + 1;
			if($left < $this->len && $this->values[$left] > $this->values[$index])
				$swap_id = $left;
			if($right < $this->len && $this->values[$right] < $this->values[$swap_id])
				$swap_id = $right;

			if($swap_id === $index) break;
			$tmp = $this->values[$index];
			$this->values[$index] = $this->values[$swap_id];
			$this->values[$swap_id] = $tmp;
			$index = $swap_id;
		}
	}

	public function enqueue($value, $priority){
		array_push($this->values, new PQNode($value, $priority));
		$this->len++;
		$this->shift_up();
	}

	public function dequeue(){
		$highest = null;
		if($this->len > 0){
			$highest = $this->values[0];
			$this->values[0] = array_pop($this->values);
			$this->len--;
			$this->shift_down();
		}
		return $highest;
	}

	public function __toString(): string{
		$os = '';
		$i = 0;
		while($i < $this->len){
			$os .= (string)$this->values[$i++];
			if($i <= $this->len - 1)
				$os .= ',';
		}
		return $os;
	}
	public function size(): int{
		return $this->len;
	}
}

if (!count(debug_backtrace(DEBUG_BACKTRACE_IGNORE_ARGS))){
	$pq = new PriorityQueue();
	$inf = 1 << 31;
	$pq->enqueue("a", 0);
	$pq->enqueue("b", $inf);
	$pq->enqueue("c", $inf);
	$pq->enqueue("d", $inf);
	$pq->enqueue("e", $inf);
	$pq->enqueue("f", $inf);
	$pq->dequeue();
	$pq->enqueue("b", 4);
	$pq->enqueue("c", 2);
	echo $pq . PHP_EOL;
	$pq->dequeue();
	$pq->enqueue("d", 4);
	echo $pq->size() . PHP_EOL;
	echo (string)$pq;
}